encodings/dict/src/variants.rs

```diff
@@ -1,26 +1,24 @@
 use vortex_array::variants::{
-    ArrayVariants, BinaryArrayTrait, BoolArrayTrait, PrimitiveArrayTrait, Utf8ArrayTrait,
+    BinaryArrayTrait, BoolArrayTrait, PrimitiveArrayTrait, Utf8ArrayTrait, VariantsVTable,
 };
-use vortex_array::ArrayDType;
-use vortex_dtype::DType;
 
-use crate::DictArray;
+use crate::{DictArray, DictEncoding};
 
-impl ArrayVariants for DictArray {
-    fn as_bool_array(&self) -> Option<&dyn BoolArrayTrait> {
-        matches!(self.dtype(), DType::Bool(..)).then_some(self)
+impl VariantsVTable<DictArray> for DictEncoding {
+    fn as_bool_array<'a>(&self, array: &'a DictArray) -> Option<&'a dyn BoolArrayTrait> {
+        Some(array)
```

Comment:
Should check the DType is bool otherwise return None; ditto the others

Reply:
No. Please read what I'm saying

Reply:
VTable implementations can make assumptions about the context in which they run because they are invoked via "entry points".

e.g. the SliceFn compute vtable assumes the bounds have already been checked by the `slice` entry-point.

These vtables can assume the DType has already been checked by `ArrayData::as_bool_array` which performs the check before delegating to the encoding to perform the downcast.